pkg/kinflate/util/util.go

```diff
@@ -0,0 +1,161 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"sort"
+
+	"github.com/ghodss/yaml"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	k8syaml "k8s.io/apimachinery/pkg/util/yaml"
+)
+
+type GroupVersionKindName struct {
+	gvk schema.GroupVersionKind
+	// name of the resource.
+	name string
+}
+
+func Decode(in []byte) (map[GroupVersionKindName]*unstructured.Unstructured, error) {
+	decoder := k8syaml.NewYAMLOrJSONDecoder(bytes.NewReader(in), 1024)
+	objs := []*unstructured.Unstructured{}
+
+	var err error
+	for {
+		var out unstructured.Unstructured
+		err = decoder.Decode(&out)
+		if err != nil {
+			break
+		}
+		objs = append(objs, &out)
+	}
+	if err != io.EOF {
+		return nil, err
+	}
+
+	m := map[GroupVersionKindName]*unstructured.Unstructured{}
+	for i := range objs {
+		metaAccessor, err := meta.Accessor(objs[i])
+		if err != nil {
+			return nil, err
+		}
+		name := metaAccessor.GetName()
+		typeAccessor, err := meta.TypeAccessor(objs[i])
+		if err != nil {
+			return nil, err
+		}
+		apiVersion := typeAccessor.GetAPIVersion()
+		kind := typeAccessor.GetKind()
+		gv, err := schema.ParseGroupVersion(apiVersion)
+		if err != nil {
+			return nil, err
+		}
+		gvk := gv.WithKind(kind)
+		gvkn := GroupVersionKindName{
+			gvk:  gvk,
+			name: name,
+		}
+		m[gvkn] = objs[i]
+	}
+	return m, nil
+}
+
+func Encode(in map[GroupVersionKindName]*unstructured.Unstructured) ([]byte, error) {
+	gvknList := []GroupVersionKindName{}
+	for gvkn := range in {
+		gvknList = append(gvknList, gvkn)
+	}
+	sort.Sort(ByGVKN(gvknList))
+
+	firstObj := true
+	var b []byte
+	buf := bytes.NewBuffer(b)
+	for _, gvkn := range gvknList {
+		obj := in[gvkn]
+		out, err := yaml.Marshal(obj)
+		if err != nil {
+			return nil, err
+		}
+		if !firstObj {
+			_, err = buf.WriteString("---\n")
+			if err != nil {
+				return nil, err
+			}
+		}
+		_, err = buf.Write(out)
+		if err != nil {
+			return nil, err
+		}
+		firstObj = false
+	}
+	return buf.Bytes(), nil
+}
+
+type mutateFunc func(interface{}) (interface{}, error)
+
+func mutateField(m map[string]interface{}, pathToField []string, createIfNotPresent bool, fns ...mutateFunc) error {
```

Comment:
This func is copied from https://github.com/kubernetes/kubectl/blob/master/pkg/kinflate/update.go#L35